### PR TITLE
go/consensus: Add block metadata to allow same-block state validation

### PR DESCRIPTION
**Pull request by kostko**
Based on #5285

### CODE REVIEW

File: go/consensus/tendermint/abci/system.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+	tmcrypto "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/crypto"
+)
+
+// prepareSystemTxs prepares a list of system transactions to be included in a proposed block in
+// case where the node is currently the block proposer.
+func (mux *abciMux) prepareSystemTxs() [][]byte {
+	var systemTxs [][]byte
+
+	// Append block metadata as a system transaction.
```
Comment (peternose) [inline go/consensus/tendermint/abci/system.go:21]: We could for loop over `SystemMethods`, do a `switch`  and put this code under `case consensus.MethodMeta`.

File: go/consensus/tendermint/abci/system.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	if len(mux.state.proposal.hash) == 0 {
+		return nil
+	}
+
+	var hasBlockMetadata bool
+	for _, tx := range mux.state.blockCtx.SystemTransactions {
+		switch tx.Method {
+		case consensus.MethodMeta:
+			// Block metadata, verify state root.
+			if hasBlockMetadata {
```
Comment (peternose) [inline go/consensus/tendermint/abci/system.go:88]: Having a counter instead of a bool flag could remove this `if`. Not sure what is better :)

**Review by peternose — APPROVED** [review]
Looks great! Left some minor comments, though.

File: go/consensus/api/meta.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
+)
+
+// MethodMeta is the method name for the special block metadata transaction.
+var MethodMeta = transaction.NewMethodName(ModuleName, "Meta", BlockMetadata{})
+
+// BlockMetadataMaxSize is the maximum size of a fully populated block metadata transaction.
+const BlockMetadataMaxSize = 10_000
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/consensus/api/meta.go]: Not really, this is needed to ensure that the block would not overflow after adding the metadata transaction so it is fine if there is some buffer. I've now also added a sanity check during system transaction creation.

**Pull request merged**